pkg/model-serving-controller/controller/model_serving_controller.go

```diff
@@ -213,6 +213,22 @@ func (c *ModelServingController) deleteModelServing(obj interface{}) {
 		}
 	}
 
+	servingGroupList, err := c.store.GetServingGroupByModelServing(utils.GetNamespaceName(mi))
+	if err != nil && !errors.Is(err, datastore.ErrServingGroupNotFound) {
+		klog.Errorf("cannot get servingGroup of modelServing: %s from map: %v", mi.GetName(), err)
+		return
+	}
+
+	for i := range servingGroupList {
+		services, err := c.getServicesByIndex(GroupNameKey, fmt.Sprintf("%s/%s", mi.Namespace, servingGroupList[i].Name))
+		if err != nil {
+			klog.Errorf("failed to get services for ServingGroup %s: %v", servingGroupList[i].Name, err)
+			return
+		}
+		// delete headless service
+		c.deleteHeadlessService(services)
```

Comment:
It is not quite correct, we may leak headless services if the modelServing is deleted when controller-manager is down.